### PR TITLE
[Core][VariableUtils] porting to parallel utilities

### PR DESCRIPTION
**Pull request by ashishdarekar**
**Description**
Porting Pure OpenMP loops to Parallel Utilities.

- ParallelUtilities
- Applications

**Changelog**
- Changed OpenMP loops from the file Variable_utils.h and Variable_utlis.cpp

### CODE REVIEW

File: kratos/utilities/variable_utils.cpp [hunk 1]
```diff
@@ -218,15 +210,11 @@ void VariableUtils::UpdateCurrentPosition(
...
-    const int num_nodes = static_cast<int>(rNodes.size());
-    const auto it_node_begin = rNodes.begin();
-
-    #pragma omp parallel for
-    for (int i_node = 0; i_node < num_nodes; ++i_node) {
-        const auto it_node  = it_node_begin + i_node;
-        const auto& r_update_coords = it_node->FastGetSolutionStepValue(rUpdateVariable, BufferPosition);
-        noalias(it_node->Coordinates()) = (it_node->GetInitialPosition()).Coordinates() + r_update_coords;
-    }
+    auto rNodes_local = rNodes;
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
    auto& rNodes_local = rNodes;
```
Pointers are automatic, references not

Comment (loumalouomega) [reply kratos/utilities/variable_utils.cpp]: Except if this is a copy...

File: kratos/utilities/variable_utils.cpp [hunk 2]
```diff
@@ -195,14 +191,10 @@ void VariableUtils::UpdateCurrentToInitialConfiguration(const ModelPart::NodesCo
...
 
-    const int num_nodes = static_cast<int>(rNodes.size());
-    const auto it_node_begin = rNodes.begin();
-
-    #pragma omp parallel for
-    for (int i=0; i<num_nodes; i++) {
-        const auto it_node  = it_node_begin + i;
-        noalias(it_node->GetInitialPosition().Coordinates()) = it_node->Coordinates();
-    }
+    auto rNodes_local = rNodes;
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
    auto& rNodes_local = rNodes;
```

File: kratos/utilities/variable_utils.cpp [hunk 3]
```diff
@@ -177,14 +177,10 @@ bool VariableUtils::CheckVariableKeys()
...
 
-    const int num_nodes = static_cast<int>(rNodes.size());
-    const auto it_node_begin = rNodes.begin();
-
-    #pragma omp parallel for
-    for (int i=0; i<num_nodes; i++) {
-        const auto it_node  = it_node_begin + i;
-        noalias(it_node->Coordinates()) = it_node->GetInitialPosition();
-    }
+    auto rNodes_local = rNodes;
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
    auto& rNodes_local = rNodes;
```

File: kratos/utilities/variable_utils.h [hunk 4]
```diff
@@ -133,13 +133,12 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
                                                         << "\n\t- Number of origin nodes: " << n_orig_nodes
                                                         << "\n\t- Number of destination nodes: " << n_dest_nodes << std::endl;
 
-        #pragma omp parallel for
-        for(int i_node = 0; i_node < n_orig_nodes; ++i_node){
-            auto it_dest_node = rDestinationModelPart.NodesBegin() + i_node;
-            const auto &it_orig_node = rOriginModelPart.NodesBegin() + i_node;
+        IndexPartition<std::size_t>(n_orig_nodes).for_each([&](std::size_t index){
+            auto it_dest_node = rDestinationModelPart.NodesBegin() + index;
+            const auto &it_orig_node = rOriginModelPart.NodesBegin() + index;
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
            const auto it_orig_node = rOriginModelPart.NodesBegin() + index;
```
In the other hand iterators don't require reference

File: kratos/utilities/variable_utils.h [hunk 5]
```diff
@@ -133,13 +133,12 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
             const auto &r_value = it_orig_node->GetSolutionStepValue(rVariable, BuffStep);
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
            const auto& r_value = it_orig_node->GetSolutionStepValue(rVariable, BuffStep);
```

File: kratos/utilities/variable_utils.h [hunk 6]
```diff
@@ -177,13 +176,12 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
             "\n\t- Number of origin nodes: " << n_orig_nodes <<
             "\n\t- Number of destination nodes: " << n_dest_nodes << std::endl;
 
-        #pragma omp parallel for
-        for(int i_node = 0; i_node < n_orig_nodes; ++i_node){
-            auto it_dest_node = rDestinationModelPart.NodesBegin() + i_node;
-            const auto &it_orig_node = rOriginModelPart.NodesBegin() + i_node;
+        IndexPartition<std::size_t>(n_orig_nodes).for_each([&](std::size_t index){
+            auto it_dest_node = rDestinationModelPart.NodesBegin() + index;
+            const auto &it_orig_node = rOriginModelPart.NodesBegin() + index;
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
            const auto it_orig_node = rOriginModelPart.NodesBegin() + index;
```

File: kratos/utilities/variable_utils.h [hunk 7]
```diff
@@ -177,13 +176,12 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
             const auto &r_value = it_orig_node->GetSolutionStepValue(rVariable, BuffStep);
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
            const auto& r_value = it_orig_node->GetSolutionStepValue(rVariable, BuffStep);
```

File: kratos/utilities/variable_utils.h [hunk 8]
```diff
@@ -626,13 +624,12 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
-        #pragma omp parallel for
-        for(int i_elems = 0; i_elems < n_orig_elems; ++i_elems){
-            auto it_dest_elems = rDestinationModelPart.ElementsBegin() + i_elems;
-            const auto &it_orig_elems = rOriginModelPart.ElementsBegin() + i_elems;
-            const auto &r_value = it_orig_elems->GetValue(rVariable);
-            it_dest_elems->SetValue(rVariable,r_value);
-        }
+        IndexPartition<std::size_t>(n_orig_elems).for_each([&](std::size_t index){
+        auto it_dest_elems = rDestinationModelPart.ElementsBegin() + index;
+        const auto &it_orig_elems = rOriginModelPart.ElementsBegin() + index;
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
        const auto it_orig_elems = rOriginModelPart.ElementsBegin() + index;
```

File: kratos/utilities/variable_utils.h [hunk 9]
```diff
@@ -626,13 +624,12 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
+        const auto &r_value = it_orig_elems->GetValue(rVariable);
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
        const auto& r_value = it_orig_elems->GetValue(rVariable);
```

**Review by loumalouomega — DISMISSED** [review]
Okay

File: kratos/utilities/variable_utils.h [hunk 10]
```diff
@@ -133,13 +133,12 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
-            auto it_dest_node = rDestinationModelPart.NodesBegin() + i_node;
-            const auto &it_orig_node = rOriginModelPart.NodesBegin() + i_node;
-            const auto &r_value = it_orig_node->GetSolutionStepValue(rVariable, BuffStep);
-            it_dest_node->GetSolutionStepValue(rDestinationVariable, BuffStep) = r_value;
-        }
+        IndexPartition<std::size_t>(n_orig_nodes).for_each([&](std::size_t index){
+            auto it_dest_node = rDestinationModelPart.NodesBegin() + index;
+            const auto it_orig_node = rOriginModelPart.NodesBegin() + index;
+            const auto& r_value = it_orig_node->GetSolutionStepValue(rVariable, BuffStep);
+            it_dest_node->FastGetSolutionStepValue(rDestinationVariable, BuffStep) = r_value;
```
Comment (philbucher) [inline kratos/utilities/variable_utils.h:140]: good catch!

**Review by philbucher — DISMISSED** [review]

**Review by philbucher — COMMENTED** [review]
@riccardotosi this ModelPart cannot be const, you modify the nodes etc
Not sure how this worked before

**Comment by riccardotosi** [comment]
> 
> 
> @riccardotosi this ModelPart cannot be const, you modify the nodes etc
> Not sure how this worked before

Me neither 🤷‍♂️

**Comment by riccardotosi** [comment]
I leave approval to @KratosMultiphysics/technical-committee.

**Comment by philbucher** [comment]
@rubenzorrilla could you please have a look over this PR to make sure everything is ok?

File: kratos/utilities/variable_utils.cpp [hunk 11]
```diff
@@ -174,35 +174,27 @@ bool VariableUtils::CheckVariableKeys()
...
-    const auto it_node_begin = rNodes.begin();
-
-    #pragma omp parallel for
-    for (int i=0; i<num_nodes; i++) {
-        const auto it_node  = it_node_begin + i;
-        noalias(it_node->Coordinates()) = it_node->GetInitialPosition();
-    }
+    BlockPartition<const ModelPart::NodesContainerType, ModelPart::NodesContainerType::const_iterator>(rNodes).for_each([](Node<3>& rNode){
+        noalias(rNode.Coordinates()) = rNode.GetInitialPosition();
+    });
```
Comment (rubenzorrilla) [inline kratos/utilities/variable_utils.cpp:183]: I'd wait for #8251 to be merged to directly pass the container.

File: kratos/utilities/variable_utils.cpp [hunk 12]
```diff
@@ -174,35 +174,27 @@ bool VariableUtils::CheckVariableKeys()
...
-    const auto it_node_begin = rNodes.begin();
-
-    #pragma omp parallel for
-    for (int i=0; i<num_nodes; i++) {
-        const auto it_node  = it_node_begin + i;
-        noalias(it_node->GetInitialPosition().Coordinates()) = it_node->Coordinates();
-    }
+    BlockPartition<const ModelPart::NodesContainerType, ModelPart::NodesContainerType::const_iterator>(rNodes).for_each([](Node<3>& rNode){
+        noalias(rNode.GetInitialPosition().Coordinates()) = rNode.Coordinates();
+    });
```
Comment (rubenzorrilla) [inline kratos/utilities/variable_utils.cpp:197]: I'd wait for #8251 to be merged to directly pass the container.

File: kratos/utilities/variable_utils.cpp [hunk 13]
```diff
@@ -211,22 +203,17 @@ void VariableUtils::UpdateInitialToCurrentConfiguration(const ModelPart::NodesCo
...
-    #pragma omp parallel for
-    for (int i_node = 0; i_node < num_nodes; ++i_node) {
-        const auto it_node  = it_node_begin + i_node;
-        const auto& r_update_coords = it_node->FastGetSolutionStepValue(rUpdateVariable, BufferPosition);
-        noalias(it_node->Coordinates()) = (it_node->GetInitialPosition()).Coordinates() + r_update_coords;
-    }
+    BlockPartition<const ModelPart::NodesContainerType, ModelPart::NodesContainerType::const_iterator>(rNodes).for_each([&rUpdateVariable, BufferPosition](Node<3>& rNode){
+        const auto& r_update_coords = rNode.FastGetSolutionStepValue(rUpdateVariable, BufferPosition);
+        noalias(rNode.Coordinates()) = (rNode.GetInitialPosition()).Coordinates() + r_update_coords;
+    });
```
Comment (rubenzorrilla) [inline kratos/utilities/variable_utils.cpp:216]: I'd wait for #8251 to be merged to directly pass the container.

File: kratos/utilities/variable_utils.cpp [hunk 14]
```diff
@@ -174,35 +174,27 @@ bool VariableUtils::CheckVariableKeys()
...
 
-    const int num_nodes = static_cast<int>(rNodes.size());
-    const auto it_node_begin = rNodes.begin();
-
-    #pragma omp parallel for
-    for (int i=0; i<num_nodes; i++) {
-        const auto it_node  = it_node_begin + i;
-        noalias(it_node->Coordinates()) = it_node->GetInitialPosition();
-    }
+    BlockPartition<const ModelPart::NodesContainerType, ModelPart::NodesContainerType::const_iterator>(rNodes).for_each([](Node<3>& rNode){
```
Comment (rubenzorrilla) [inline kratos/utilities/variable_utils.cpp]: I think that the changes in #8251 make possible to use the short syntax. I'd wait for it to be merged.

File: kratos/utilities/variable_utils.cpp [hunk 15]
```diff
@@ -174,35 +174,27 @@ bool VariableUtils::CheckVariableKeys()
...
 
-    const int num_nodes = static_cast<int>(rNodes.size());
-    const auto it_node_begin = rNodes.begin();
-
-    #pragma omp parallel for
-    for (int i=0; i<num_nodes; i++) {
-        const auto it_node  = it_node_begin + i;
-        noalias(it_node->GetInitialPosition().Coordinates()) = it_node->Coordinates();
-    }
+    BlockPartition<const ModelPart::NodesContainerType, ModelPart::NodesContainerType::const_iterator>(rNodes).for_each([](Node<3>& rNode){
```
Comment (rubenzorrilla) [inline kratos/utilities/variable_utils.cpp]: Same.

File: kratos/utilities/variable_utils.cpp [hunk 16]
```diff
@@ -211,22 +203,17 @@ void VariableUtils::UpdateInitialToCurrentConfiguration(const ModelPart::NodesCo
...
-    const int num_nodes = static_cast<int>(rNodes.size());
-    const auto it_node_begin = rNodes.begin();
-
-    #pragma omp parallel for
-    for (int i_node = 0; i_node < num_nodes; ++i_node) {
-        const auto it_node  = it_node_begin + i_node;
-        const auto& r_update_coords = it_node->FastGetSolutionStepValue(rUpdateVariable, BufferPosition);
-        noalias(it_node->Coordinates()) = (it_node->GetInitialPosition()).Coordinates() + r_update_coords;
-    }
+    BlockPartition<const ModelPart::NodesContainerType, ModelPart::NodesContainerType::const_iterator>(rNodes).for_each([&rUpdateVariable, BufferPosition](Node<3>& rNode){
```
Comment (rubenzorrilla) [inline kratos/utilities/variable_utils.cpp]: Same.

**Review by rubenzorrilla — COMMENTED** [review]
Changes make sense to me. Thanks for the clean up. No comments aside of the #8251 - related ones.

**Comment by rubenzorrilla** [comment]
BTW, I think we should stick to the style guide in the lambdas too (e.g. `index` should be `Index` when using the `IndexPartitioner`). I didn't want to be that picky (feel free to dismiss this).

File: kratos/utilities/variable_utils.cpp (hunk 14 above)
Comment (ashishdarekar) [reply kratos/utilities/variable_utils.cpp]: Done

**Review by philbucher — APPROVED** [review]
👍

**Review by rubenzorrilla — APPROVED** [review]

**Pull request merged**